src/time.cpp

```diff
@@ -69,6 +71,27 @@ namespace g3 {
    /// * format string must conform to std::put_time
    /// This is similar to std::put_time(std::localtime(std::time_t*), time_format.c_str());
 
+   std::string localtime_formatted(const timespec &time_snapshot, const std::string &time_format) {
+      ushort fcount = 0;
+      for(auto found_ptr = time_format.c_str(); (found_ptr = strstr(found_ptr, "%f")) != nullptr; found_ptr++, fcount++);
+      const size_t format_size = time_format.length() + fcount * 7 + 1;  // len(format) + num(nanosec size - len("%f")) + zero
+
+      // creating a new buffer on the stack conaining format string
+      char format_buffer[format_size];
+      strcpy(format_buffer, time_format.c_str());
+
+      for(auto found_ptr = format_buffer; (found_ptr = strstr(found_ptr, "%f")) != nullptr; found_ptr++) {
+         strcpy(found_ptr + 9, found_ptr + 2);
+         memcpy(found_ptr, "000000000", 9);
+         auto nsec_str = std::to_string(time_snapshot.tv_nsec);
+         auto nsec_len = strlen(nsec_str.c_str());
+         memcpy(found_ptr, nsec_str.c_str() + nsec_len - 9, nsec_len);
+      }
```

Comment:
This all looks correct but it can be simplified a great deal. Something like this should work:

```
std::string localtime_formatted(const timespec &time_snapshot, const std::string &time_format) {
   auto format_buffer = time_format;
   auto pos = format_buffer.find(kNanoSecondIdentifier);
   while (pos != std::string::npos) {
   	  format_buffer.replace(pos, kNanoSecondIdentifier.size(), ns);
   	  pos = format_buffer.find(kNanoSecondIdentifier, pos + ns.size());
   }
      std::tm t = localtime(time_snapshot.tv_sec);
      return g3::internal::put_time(&t, format_buffer);
}
```

ref: simple testing of this at: <script src="http://ideone.com/e.js/lSbs1Z" type="text/javascript" ></script>